### PR TITLE
DBZ-588 Fail on inconsistent schema

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-588

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 1]
```diff
@@ -638,7 +640,48 @@ protected void handleUpdateTableMetadata(Event event) {
...
+                            "Use the mysqlbinlog tool to view the problematic event: mysqlbinlog --start-position={} --stop-position={} --verbose {}",
+                            source.offset(),
+                            System.lineSeparator(),
+                            System.lineSeparator(),
+                            eventHeader.getPosition(),
+                            eventHeader.getNextPosition(),
+                            source.binlogFilename()
+                    );
+                }
+            } else {
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Let's use the "no keywords after closing brace" style as discussed today.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 2]
```diff
@@ -638,7 +640,48 @@ protected void handleUpdateTableMetadata(Event event) {
         if (recordMakers.assign(tableNumber, tableId)) {
             logger.debug("Received update table metadata event: {}", event);
         } else {
-            logger.debug("Skipping update table metadata event: {}", event);
+            if (context.dbSchema().isTableMonitored(tableId)) {
+                EventHeaderV4 eventHeader = event.getHeader();
+
+                if (inconsistentSchemaHandlingMode == EventProcessingFailureHandlingMode.FAIL) {
+                    logger.error(
+                            "Schema inconsistency detected while processing binlog event at offset {}.{}" +
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: "Schema inconsistency" sounds a bit vague. How about:

"Encountered change event for table <xyz> whose schema isn't known to this connector. One possible cause is an incomplete database history topic. Take a new snapshot in this case".

Also the event itself isn't logged anymore, I think we should keep doing that.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 3]
```diff
@@ -638,7 +640,48 @@ protected void handleUpdateTableMetadata(Event event) {
...
+                            System.lineSeparator(),
+                            eventHeader.getPosition(),
+                            eventHeader.getNextPosition(),
+                            source.binlogFilename()
+                    );
+                    throw new ConnectException("Inconsistency in internal schema detected");
+                } else if (inconsistentSchemaHandlingMode == EventProcessingFailureHandlingMode.WARN) {
+                    logger.warn(
+                            "Schema inconsistency detected while processing binlog event at offset {}.{}" +
+                            "This exception will be ignored and the event be skipped.{}" +
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Technically, there's no exception.

"Encountered change event for table whose schema isn't known to this connector. One possible cause is an incomplete database history topic. Take a new snapshot in this case. The event will be ignored".

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 4]
```diff
@@ -638,7 +640,48 @@ protected void handleUpdateTableMetadata(Event event) {
...
+                            System.lineSeparator(),
+                            System.lineSeparator(),
+                            eventHeader.getPosition(),
+                            eventHeader.getNextPosition(),
+                            source.binlogFilename()
+                    );
+                } else {
+                    logger.debug(
+                            "Schema inconsistency detected while processing binlog event at offset {}.{}" +
+                            "This exception will be ignored and the event be skipped.{}" +
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: As above.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/BinlogReaderIT.java [hunk 5]
```diff
@@ -351,6 +357,63 @@ public void shouldHandleMySQLTimeCorrectly() throws Exception {
...
+
+        // Poll for records ...
+        // Testing.Print.enable();
+        int expected = 9 + 9 + 4 + 5; // only the inserts for our 4 tables in this database
+        int consumed = consumeAtLeast(expected);
+        assertThat(consumed).isGreaterThanOrEqualTo(expected);
+
+        reader.stop();
+        reader.start();
+        reader.context.dbSchema().applyDdl(context.source(), DATABASE.getDatabaseName(), "DROP TABLE customers", null);
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/BinlogReaderIT.java]: Nice, I was wondering how you'd test it.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/BinlogReaderIT.java [hunk 6]
```diff
@@ -351,6 +357,63 @@ public void shouldHandleMySQLTimeCorrectly() throws Exception {
...
+        try (
+                final MySQLConnection db = MySQLConnection.forTestDatabase(DATABASE.getDatabaseName());
+                final JdbcConnection connection = db.connect();
+                final Connection jdbc = connection.connection();
+                final Statement statement = jdbc.createStatement()) {
+            statement.executeUpdate("INSERT INTO customers VALUES (default,'John','Lazy','john.lazy@acme.com')");
+        }
+    }
+
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/BinlogReaderIT.java]: Just one new line.

**Comment by jpechane** [comment]
@gunnarmorling Wording changed

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 7]
```diff
@@ -638,7 +640,54 @@ protected void handleUpdateTableMetadata(Event event) {
...
+                            eventHeader.getPosition(),
+                            eventHeader.getNextPosition(),
+                            source.binlogFilename()
+                    );
+                    throw new ConnectException("Encountered change event for table whose schema isn't known to this connector");
+                }
+                else if (inconsistentSchemaHandlingMode == EventProcessingFailureHandlingMode.WARN) {
+                    logger.warn(
+                            "Encountered change event '{}' at offset {} for table whose schema isn't known to this connector. One possible cause is an incomplete database history topic. Take a new snapshot in this case.{}" +
+                            "The event will be ingored.{}" +
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: s/ingored/ignored/

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java [hunk 8]
```diff
@@ -638,7 +640,54 @@ protected void handleUpdateTableMetadata(Event event) {
...
+                            System.lineSeparator(),
+                            eventHeader.getPosition(),
+                            eventHeader.getNextPosition(),
+                            source.binlogFilename()
+                    );
+                }
+                else {
+                    logger.debug(
+                            "Encountered change event '{}' at offset {} for table whose schema isn't known to this connector. One possible cause is an incomplete database history topic. Take a new snapshot in this case.{}" +
+                            "The event will be ingored.{}" +
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java]: Same here.

**Comment by jpechane** [comment]
If I understand it correctly then evey change consists of two events - table metadata and the event itself. If we handle table metadata as in this PR then the only case when converter is null is when the table is filtered out. Hence no change in the code. If you wish we can take a defensive approach and double check in change events too.

**Comment by gunnarmorling** [comment]
Yeah, I think let's better be safe than sorry.

**Comment by gunnarmorling** [comment]
Thanks, @jpechane. Rebased and applied. Could you also add a doc PR for the new option?

**Pull request closed without merging**